### DISCUSSION TITLE
From weppy to Emmett

### DISCUSSION THREAD

**Discussion by josejachuf** [emmett-framework/emmett#474] [Q&A] [answered]
Migrating a Weppy app to Emmett (2.5.4) happens that in the following consultation all virtual fields (rowattr and rowmethod) do not work

   ```
 reparticiones = [r.id for r in session.auth.user.reparticiones()]
    qry = (db.Tramite.a_visar.belongs([1, 2]))
    qry &= (db.Tramite.estado != 5)
    qry &= (db.Tramite.borrado == False)
    qry &= (db.TramiteReparticion.reparticion.belongs(reparticiones))

    ...
    rows = db(qry).select(
        db.Tramite.id,
        db.Tramite.updated_at,
        db.Tramite.tipo,
        db.Tramite.arquitecto,
        db.Tramite.comitente_nombre,
        db.Tramite.estado,
        db.Tramite.area,
        db.Tramite.a_visar,
        db.Tramite.expediente_municipal,
        left=[db.TramiteReparticion.on(
            db.Tramite.id == db.TramiteReparticion.tramite)
        ],
        orderby=~db.Tramite.updated_at,
        paginate=(int(page), page_size),
        distinct=True,
    )
```

```
<td><b>{{=row.tramite}}</b></td>
AttributeError: tramite
```
tramite is a **rowattr**

when query is directly on the object if it works fine, for example:

```
rows = Tramite.my_scope().select(
        orderby=~db.Tramite.updated_at,
        paginate=(int(page), page_size)
        )
```

**Comment by gi0baro** [comment, marked as answer]
@josejachuf in Emmett virtual fields are applied only on "concrete" records, which means only for cases where you select all the fields of a table.

This is due to the fact the ORM can't guess in advance which fields are required in the record to make virtual ones working.

**Comment by josejachuf** [comment]
I found that if I do not give the fields as the select argument it works fine

This works fine
```
rows = Tramite.my_scope().select(
        orderby=~db.Tramite.updated_at,
        paginate=(int(page), page_size)
        )
```

This not works:
```
rows = Tramite.my_scope().select(
        Tramite.id,
        orderby=~db.Tramite.updated_at,
        paginate=(int(page), page_size)
        )
```

**Comment by josejachuf** [comment]
Thanks @gi0baro 
In "Tramite" I have several JSON fields.

When I do the select of all the fields in the Shell it works, but when I am running it gives me this mistake.

> rv = self.cursor.execute(command, *args[1:], **kwargs)
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> psycopg2.errors.UndefinedFunction: could not identify an equality operator for type json
> LINE 1: ...io", "tramites"."latitud", "tramites"."longitud", "tramites"...

If I enter the fields, the Virtual Fields does not work and if I put all I get error with the JSON

Any ideas?

**Comment by josejachuf** [comment]
The problem is in the distinct with the JSON fields. The error happens if I run directly on the PostgreSql console
